pkg/operator/missingstaticpodcontroller/missingstaticpodcontroller.go

```diff
@@ -0,0 +1,209 @@
+	for node, latestInstallerPodOnNode := range latestInstallerPodsByNode {
+		installerPodRevision, err := getInstallerPodRevision(latestInstallerPodOnNode)
+		if err != nil {
+			// we expect every installer pod to have a installerPodRevision in its name, unexpected error here
+			return fmt.Errorf("failed to get installerPodRevision for installer pod %q - %w", latestInstallerPodOnNode.Name, err)
+		}
+
+		finishedAt, ok := installerPodFinishedAt(latestInstallerPodOnNode)
+		if !ok {
+			// either it's in the process of running, or it ran into an error
+			continue
+		}
+		threshold := terminationGracePeriodSeconds(latestInstallerPodOnNode) + 10*time.Second
+		staticPodRevisionOnThisNode := getStaticPodCurrentRevision(node, originalOperatorStatus)
+		if time.Since(finishedAt) > threshold &&
+			staticPodRevisionOnThisNode < installerPodRevision {
+			// if we are here:
+			//  a: the latest installer pod successfully completed at finishedAt
+			//  b: it has been more than 'terminationGracePeriodSeconds' + 10s since the
+			//     installer pod has completed at finishedAt
+			//  c. the static pod is not at correct installerPodRevision yet
+			// then all of the above conditions are true, so we should go degraded
+			//
+			// TODO: set a degraded condition
+			syncCtx.Recorder().Eventf("MissingStaticPod", "static pod lifecycle failure - installer pod: %q with revision: %d completed at: %s, static pod revision: %d",
+				latestInstallerPodOnNode.Name, installerPodRevision, finishedAt, staticPodRevisionOnThisNode)
+		}
+	}
+
+	return nil
+}
+
+func getStaticPodCurrentRevision(node string, status *operatorv1.StaticPodOperatorStatus) int {
+	var nodeStatus *operatorv1.NodeStatus
+	for i := range status.NodeStatuses {
+		if status.NodeStatuses[i].NodeName == node {
+			nodeStatus = &status.NodeStatuses[i]
+			break
+		}
+	}
+
+	if nodeStatus == nil {
+		return 0
+	}
+	return int(nodeStatus.CurrentRevision)
+}
+
+func terminationGracePeriodSeconds(pod *corev1.Pod) time.Duration {
+	value := pod.Spec.TerminationGracePeriodSeconds
+	if value == nil {
+		return 0
+	}
+
+	return time.Duration(*value * int64(time.Second))
+}
+
+func getMostRecentInstallerPodByNode(pods []*corev1.Pod) map[string]*corev1.Pod {
+	mostRecentInstallerPodByNode := map[string]*corev1.Pod{}
+	byNodes := getInstallerPodsByNode(pods)
+	for node, installerPodsOnThisNode := range byNodes {
+		if len(installerPodsOnThisNode) == 0 {
+			continue
+		}
+		sort.Sort(byRevision(installerPodsOnThisNode))
+		mostRecentInstallerPodByNode[node] = installerPodsOnThisNode[len(installerPodsOnThisNode)-1]
+	}
+
+	return mostRecentInstallerPodByNode
+}
+
+func getInstallerPodsByNode(pods []*corev1.Pod) map[string][]*corev1.Pod {
+	byNodes := map[string][]*corev1.Pod{}
+	for i := range pods {
+		pod := pods[i]
+		if !strings.HasPrefix(pod.Name, "installer-") {
+			continue
+		}
+
+		nodeName := pod.Spec.NodeName
+		if len(nodeName) == 0 {
```

Comment:
technically this should never happen, not sure if we would like to return an error just in case